modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
-  // implementation for assignment will become necessary.
-  inline proc =(ref a: string, b: string) { __primitive("move", a, b); }
+      return this.buff:c_string;
+    }
 
-  inline proc ==(a: string, b: string) return (__primitive("string_compare", a.c_str(), b.c_str()) == 0);
-  inline proc !=(a: string, b: string) return (__primitive("string_compare", a.c_str(), b.c_str()) != 0);
+    inline proc length return len;
+    //inline proc size return len;
+
+    // Localize gets a version of this that is on the currently executing
+    // locale. If this is already on the current locale we return a shallow
+    // copy, otherwise a remote copy is performed.
+    inline proc localize() /*: string*/ {
+        const ret: string = if this.locale.id != chpl_nodeID
+          then this // assignment makes it local
+          else new string(this, owned=false);
+        return ret;
+    }
 
 
-  inline proc <=(a: string, b: string) return a.c_str()<=b.c_str();
-  inline proc >=(a: string, b: string) return a.c_str()>=b.c_str();
-  inline proc <(a: string, b: string) return a.c_str()<b.c_str();
-  inline proc >(a: string, b: string) return a.c_str()>b.c_str();
+    // Steals ownership of string.buff. This is unsafe and you probably don't
+    // want to use it. I'd like to figure out how to get rid of it entirely.
+    proc _steal_buffer() /*: bufferType*/ {
+      const ret = this.buff;
+      this.buff = nil;
+      this.len = 0;
+      this._size = 0;
+      return ret;
+    }
 
-  //
-  // primitive string functions and methods
-  //
-  inline proc ascii(a: string) return ascii(a.c_str());
-  inline proc string.length return this.c_str().length;
-  inline proc string.size return this.length;
-  inline proc string.substring(i: int) {
-    var cs = this.c_str().substring(i);
-    const ret = toString(cs);
-    // toString steals the returned c_string_copy, so no need to free it.
-    return ret;
-  }
-  inline proc string.substring(r: range(?)) {
-    var cs = this.c_str().substring(r);
-    const ret = toString(cs);
-    // toString steals the returned c_string_copy, so no need to free it.
+    // TODO: cant explicitly state return type right now due to a bug in the
+    // compiler. Uncomment this funciton and others when possible.
+    iter these() /*: string*/ {
+      for i in 1..this.len {
+        yield this[i];
+      }
+    }
+
+    proc this(i: int) /*: string*/ {
+      if i <= 0 || i > this.len then halt("index out of bounds of string");
+
+      var ret: string;
+      const newSize = chpl_mem_goodAllocSize(2);
+      ret._size = max(chpl_string_min_alloc_size, newSize.safeCast(int));
+      ret.len = 1;
+      ret.buff = chpl_mem_alloc(ret._size.safeCast(size_t),
+                                CHPL_RT_MD_STR_COPY_DATA): bufferType;
+      ret.owned = true;
+
+      const remoteThis = this.locale.id != chpl_nodeID;
+      if remoteThis {
+        chpl_string_comm_get(ret.buff, this.locale.id, this.buff, this.len.safeCast(size_t));
```

Comment:
I'm pretty sure this is wrong. And will lead to memory corruption. The comm_get call here is copying the entire string in this.len into ret.buff, but we didn't appropriately size ret.buff for that, and we probably only want to copy 1 character anyway.